### PR TITLE
image-pruner: prune images in their own jobs

### PR DESCRIPTION
**Pull request by miminar**
Instead of pruning in phases:

    all streams -> all layers -> all blobs -> manifests -> images

Prune individual images in parallel jobs:

    all streams -> parallel [
       image1's layers -> image1's blobs -> ... -> image1,
       image2's layers -> image2's blobs -> ... -> image2,
       ...
    ]

A failure in streams prune phase is not fatal anymore.

Resolves: [rhbz#1567657](https://bugzilla.redhat.com/show_bug.cgi?id=1567657)

Additionally, previously manifest blobs weren't removed from the blob store. This PR removes the manifests of deleted images from blob store as well.

**Remarks:**
- parallel execution makes it a bit more complicated
- in order to make it more legible and reviewable, the parallelism with the handling of blocked images could be moved to a follow-up PR
- this doesn't aim to make the image pruning 100% safe - it just attempts to make it error tolerable while maintaining as much consistency as possible

**TODOs:**
- [x] - pretty print result summary
- [x] - avoid potential races caused by using shared API clients by intantiating them separately for each worker
- [x] - resolve races in tests where the counters are shared between go-routines
- [x] - detect changes to image streams and update graph accordingly on the fly
- [x] - and image creations/deletions and update graph accordingly

### CODE REVIEW

**Comment by michojel** [comment]
Races expected...

/test 

Publishing this in early state to gather some feedback.

@dmage, @legionus, @bparees, @coreydaley PTAL

**Comment by miminar** [comment]
Publishing this in early state to gather some feedback.

@dmage, @legionus, @bparees, @coreydaley PTAL

**Comment by legionus** [comment]
Dynamically updated graph will not help to achieve greater consistency. You made a very comlicate code and it will speed up the work of the prunner, but this does not help with maintaining the integrity of the database. Removing an image is not atomic and in case of an error you will break the database.

In connection, I believe that this approach is not effective. I would suggest a different approach to prunning: `mark-and-sweep`. At the first iteration we mark objects with an annotation with the date of the planned deletion (for example, during 3 days). At the next iteration, we delete the objects whose label still stands. In this case, the graph is not updated dynamically.

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 1]
```diff
@@ -56,8 +54,24 @@ const (
...
 	// ImageNode to an ImageComponentNode.
 	ReferencedImageLayerEdgeKind = "ReferencedImageLayer"
+
+	// ReferencedImageManifestEdgeKind defines an edge from an ImageStreamNode or an
+	// ImageNode to an ImageComponentNode.
+	ReferencedImageManifestEdgeKind = "ReferencedImageManifest"
+
+	// UnreferencedImageComponentEdgeKind is an edge from an ImageNode to an ImageComponentNode denoting that
+	// the component is currently being unreferenced in a running job.
+	UnreferencedImageComponentEdgeKind = "UnreferencedImageComponentToDelete"
```
Comment (dmage) [inline pkg/oc/admin/prune/imageprune/prune.go]: This feels very artificial. Why do we need these "negative" references?

Can we structure the pruner such a way that we remove only nodes which don't have any references?

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 2]
```diff
@@ -914,13 +948,39 @@ func pruneStreams(
...
+			deletions = append(deletions, Deletion{Node: streamNode})
 		}
 	}
 
 	glog.V(4).Infof("Done removing pruned image references from streams")
-	return nil
+	return
+}
+
+func strenghtenReferencesFromFailedImageStreams(g genericgraph.Graph, failures []Failure) {
```
Comment (dmage) [inline pkg/oc/admin/prune/imageprune/prune.go]: s/strenghten/strengthen/

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 3]
```diff
@@ -398,10 +422,18 @@ func (p *pruner) addImageStreamsToGraph(streams *imageapi.ImageStreamList, limit
...
-					} else {
+						break
+					case imagegraph.ImageComponentTypeLayer:
 						p.g.AddEdge(imageStreamNode, s, ReferencedImageLayerEdgeKind)
+						break
+					case imagegraph.ImageComponentTypeManifest:
+						p.g.AddEdge(imageStreamNode, s, ReferencedImageManifestEdgeKind)
+						break
+					default:
+						panic(fmt.Sprintf("unhandeled image component type %q", cn.Type))
```
Comment (dmage) [inline pkg/oc/admin/prune/imageprune/prune.go]: Can we return this error?

File: pkg/oc/admin/prune/imageprune/worker.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+		}
+		out <- *w.prune(job)
+	}
+}
+
+func (w *worker) prune(job *Job) *JobResult {
+	res := &JobResult{Job: job}
+
+	// If namespace is specified prune only ImageStreams and nothing more.
+	if len(w.algorithm.namespace) > 0 {
```
Comment (dmage) [inline pkg/oc/admin/prune/imageprune/worker.go]: When may this condition become true?

**Comment by michojel** [comment]
@legionus I don't see how `mark-and-sweep` approach makes this safer and simpler. What prevents users from making new references to marked images during the sweep phase?
I know we discussed this before and could not agree on anything 100% bullet proof even when using locks in etcd. And introducing locking mechanism for the sake of safe image pruning will make the subject even more complex.

The aim of this PR is not to make the pruning safer. The purpose is to tolerate errors that prevent customers from pruning anything by doing the pruning iteratively - image after image which brings a vague atomicity to the pruning.

I don't think this change prevents us from making the pruning safer in the future. It does not prevents us from implementing `mark-and-sweep`, etcd or storage locking. All it does is just introduction of job-based pruning while building on top of what we already have.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 3 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Well, I don't expect this to ever fire. But sure.

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 5]
```diff
@@ -982,170 +1042,430 @@ func tagEventIsPrunable(
...
+				}
+				break
+			}
+			p.unreferenceComponents(job)
+			jobChan <- job
 		}
-	}
 
-	return true
+		select {
```
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: It's supposed to handle events either from workers, image streams listener or image listener (which are not there yet).

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 6]
```diff
@@ -982,170 +1042,430 @@ func tagEventIsPrunable(
...
+			}
+			edge := p.g.Edge(isn, d.Node)
+			if edge == nil {
+				continue
+			}
+			p.g.RemoveEdge(edge)
+		case *imagegraph.ImageStreamNode:
+			// ignore
+		default:
+			panic(fmt.Sprintf("unhandeled graph node %t", d.Node))
```
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: I can return the error for sure. This was meant rather as a debug statement until all the types got handled.

File: pkg/oc/admin/prune/imageprune/worker.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+func (w *worker) Run(in <-chan *Job, out chan<- JobResult) {
+	for {
+		job := <-in
+		if job == nil {
+			return
+		}
+		out <- *w.prune(job)
+	}
```
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: Good idea, I'll check that out.

File: pkg/oc/admin/prune/imageprune/worker.go [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+			w.registryClient,
+			w.registryURL,
+			job.Components,
+			w.manifestPruner,
+		))
+	}
+
+	if len(res.Failures) > 0 {
+		// TODO: include image as a failure as well for the sake of summary's completness
+		return res
```
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: True, I'm keeping the current behaviour that preserves image objects for future prunes so that blobs that failed to be pruned are recognized on the subsequent prunes. If we remove the image, we loose the option to prune its blobs using this kind of pruner.
But since we have a hard-pruner now, we could do the removal. I just hesitate to make it a default behaviour. Maybe a flag `--keep-broken-images` could be useful here. WDYT?

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 9]
```diff
@@ -333,6 +356,7 @@ func (p *pruner) addImagesToGraph(images *imageapi.ImageList) []error {
 //
 // addImageStreamsToGraph also adds references from each stream to all the
 // layers it references (via each image a stream references).
+// TODO: identify streams with non-existing images for later cleanup
 func (p *pruner) addImageStreamsToGraph(streams *imageapi.ImageStreamList, limits map[string][]*kapi.LimitRange) []error {
```
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Good catch. Haven't even realized that. I'm more leaning towards the panics for errors not aimed at the end user but a developer (saying, you're code is buggy, go fix it) such as this one. But I don't feel that strong about it, so I'll look into returning it as an error.

File: pkg/oc/admin/prune/imageprune/worker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: A scenario speaking against the image deletion regardless of prio failure is that the registry lost write permission to the storage and deletion of each blob fails (already happened to some customers). In this case all images are deleted but all the blobs remain on the storage.
Is it worse than broken images? Hard to tell. It depends on what the customer wants to achieve - either he's running out of storage space or etcd is too big and slow or even both.
Therefor I'm more inclined to having a flag like `--keep-broken-images-on-failure` defaulting to `true` (current behaviour) so that the customer can choose what he prefers.

@bparees thoughts?

[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: I'm just afraid that hard-prune is not a very popular solution to this problem and am reluctant to make it a mandatory post-routine.

[the remark this answers is not in the document]
Comment (bparees) [reply pkg/oc/admin/prune/imageprune/worker.go]: it won't be mandatory unless there are other issues in their registry.  if we can't delete layers due to storage issues, they're going to have to fix their storage issues anyway... running hard prune again after doing so seems like a small additional burden.

Is there some other case where you think it would be better to leave the image data in place if we cannot remove the blob data?

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 1 above)
Comment (bparees) [reply pkg/oc/admin/prune/imageprune/prune.go]: +1 on tracking this elsewhere.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 3 above)
Comment (bparees) [reply pkg/oc/admin/prune/imageprune/prune.go]: also should be "unhandled"

but yeah, i don't think we ever want to panic, even in the cli.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 2 above)
Comment (bparees) [inline pkg/oc/admin/prune/imageprune/prune.go]: needs godoc

Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: You're not going to believe it, but my [offline dictionary](https://cs.glosbe.com/source/2721) contains both :smile:. I cannot find it online though so I'm going to trust you.

File: pkg/oc/admin/prune/imageprune/worker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (legionus) [reply pkg/oc/admin/prune/imageprune/worker.go]: > In this case all images are deleted but all the blobs remain on the storage.
> Is it worse than broken images?

@miminar No, broken images mush worse. We have few ways how to remove blobs completely or even restore images in the etcd, but we do not have tools to restore blobs in the storage to fix images. We have only diagnostic tool  which can help to find such broken images. User should remove image from etcd and re-push such image by hands to fix them. Until he does this, these  images will be broken. There will be errors in push and pull. It seems to me that this is a much worse scenario, because the cluster goes into an inconsistent state and the correction requires manual intervention.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 1 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Reworked and simplified a bit.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 2 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Fixed

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 3 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Fixed.

File: pkg/oc/admin/prune/imageprune/worker.go (hunk 4 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: Good catch, the statement was misplaced. It's removed now.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Error reported now without a panic.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Not panicking any more.

File: pkg/oc/admin/prune/imageprune/worker.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: > What about keeping the image only if there were no successful blob deletions?

Went for this one. Please let me know if you have any objections.

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 2 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: Godoc'd

**Comment by miminar** [comment]
Rebased and added imagestream event handling.

**Comment by miminar** [comment]
Added image event handling.

**Comment by miminar** [comment]
Regarding the following error:

<details><p>

```
panic: concrete: node ID collision: 272
    
    goroutine 1 [running]:
    github.com/openshift/origin/vendor/github.com/gonum/graph/concrete.(*DirectedGraph).AddNode(0xc420ffdbf0, 0x4444de0, 0xc42157a0c0)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/vendor/github.com/gonum/graph/concrete/directed.go:67 +0x3df
    github.com/openshift/origin/pkg/oc/graph/genericgraph.uniqueNamedGraph.FindOrCreate(0x447b280, 0xc420ffdbf0, 0xc420ffdc20, 0xc421311400, 0x4e, 0xc4215412a0, 0x7f0677497000, 0x0, 0x25a4960)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/pkg/oc/graph/genericgraph/graph.go:562 +0x127
    github.com/openshift/origin/pkg/oc/graph/genericgraph.(*Graph).FindOrCreate(0xc421546380, 0xc421311400, 0x4e, 0xc4215412a0, 0x2a3c0a0, 0xc421311400, 0x7f0677359128)
    	<autogenerated>:1 +0x6c
    github.com/openshift/origin/pkg/oc/graph/genericgraph.EnsureUnique(0x7f0677359128, 0xc421546380, 0xc421311400, 0x4e, 0xc4215412a0, 0xc421546380, 0xc4210169a0)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/pkg/oc/graph/genericgraph/graph.go:65 +0x4f
    github.com/openshift/origin/pkg/oc/graph/imagegraph/nodes.EnsureImageNode(0x4483bc0, 0xc421546380, 0xc42162e000, 0xc421546380, 0x1)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/pkg/oc/graph/imagegraph/nodes/nodes.go:13 +0xde
    github.com/openshift/origin/pkg/oc/admin/prune/imageprune.(*pruner).addImagesToGraph(0xc4210169a0, 0xc421181198, 0x20, 0xc421181178, 0x1)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/pkg/oc/admin/prune/imageprune/prune.go:348 +0x532
    github.com/openshift/origin/pkg/oc/admin/prune/imageprune.(*pruner).handleImageEvent(0xc4210169a0, 0xc421554678, 0x5, 0x4452d20, 0xc42156d400)
    	/tmp/openshift/build-rpms/rpm/BUILD/origin-3.10.0/_output/local/go/src/github.com/openshift/origin/pkg/oc/admin/prune/imageprune/prune.go:868 +0x41b
```

</p></details>

Apparently, there's a bug in the gonum/graph package we use preventing us from adding a new node after deletion (the new node receives ID that already exists). I'm now bumping its version to the k8s level and is if it got fixed.

**Comment by miminar** [comment]
Fixed the [above bug](https://github.com/openshift/origin/pull/19468#issuecomment-386611771) by bumping gonum/graph package and rebased.

File: pkg/oc/graph/imagegraph/nodes/nodes_test.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	node = FindImage(g, img1.Name)
+	if node == nil {
+		t.Fatalf("expected node, not nil")
+	}
+	if node.Image != &img1 {
+		t.Fatalf("got unexpected image: %s != %s", node.Image.Name, img1.Name)
+	}
+	/**
+	 * NOTE: it looks like removed node cannot be re-added
+	if len(g.Nodes()) != 2 {
```
Comment (miminar) [inline pkg/oc/graph/imagegraph/nodes/nodes_test.go]: While adding this test I found another limitation of the graph algorithm. The consequence is that image or image stream that is re-added after being pruned will be ignored.

I'd say that the odds for the same image to be added while being pruned is pretty low but it's still unfortunate.

File: pkg/oc/graph/appsgraph/edge_test.go [hunk 11]
```diff
@@ -41,7 +41,7 @@ func TestNamespaceEdgeMatching(t *testing.T) {
 	fn("other", g)
 	AddAllDeploymentConfigsDeploymentEdges(g)
 
-	if len(g.Edges()) != 4 {
+	if len(g.Edges()) != 6 {
```
Comment (miminar) [inline pkg/oc/graph/appsgraph/edge_test.go]: Previously, two edges between the same nodes with opposite orientation were returned as one edge (even though this is a directed graph). Looks like this a result of an upstream fix.

**Comment by miminar** [comment]
Ready for review.

File: pkg/oc/graph/genericgraph/graph.go [hunk 12]
```diff
@@ -365,21 +342,21 @@ func (g Graph) AddEdge(from, to graph.Node, edgeKind string) {
...
-		case concrete.WeightedEdge:
-			if fn(g, t.From(), t.To(), t.Edge.(Edge).Kinds()) {
-				g.internal.SetEdge(t.Edge.(Edge), t.Cost)
-			}
 		case Edge:
 			if fn(g, t.From(), t.To(), t.Kinds()) {
-				g.internal.SetEdge(t, 1.0)
+				g.internal.SetEdge(t)
+			}
+		case simple.Edge:
```
Comment (dmage) [inline pkg/oc/graph/genericgraph/graph.go]: It seems that before changes this function was able to handle only genericgraph.Edge. Is this case ever executed?

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 13]
```diff
@@ -1011,170 +1132,385 @@ func tagEventIsPrunable(
...
-			glog.V(4).Infof("Config %v has an image predecessor", cn)
-			return false
+// runLoop processes the queue of prunable images until empty. It makes the workers busy and updates the graph
+// with each change.
+func (p *pruner) runLoop(
+	jobChan chan<- *Job,
+	resultChan <-chan JobResult,
+) (deletions []Deletion, failures []Failure) {
+	imgUpdateChan := p.imageWatcher.ResultChan()
+	isUpdateChan := p.imageStreamWatcher.ResultChan()
```
Comment (bparees) [inline pkg/oc/admin/prune/imageprune/prune.go]: what's the purpose of adding these watches?  Trying to handle the case that a reference to a layer is added to an image or imagestream while we're in the middle of pruning?  It seems unrelated to the fundamental purpose of the PR (to parallelize the image pruning operations)

File: pkg/oc/admin/prune/imageprune/worker.go [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+
+	"github.com/golang/glog"
+	gonum "github.com/gonum/graph"
+
+	kerrapi "k8s.io/apimachinery/pkg/api/errors"
+
+	imagegraph "github.com/openshift/origin/pkg/oc/graph/imagegraph/nodes"
+)
+
+// ComponentRetention knows all the places where image componenet needs to be pruned (e.g. global blob store
```
Comment (bparees) [inline pkg/oc/admin/prune/imageprune/worker.go]: s/componenet/component/

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 13 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/prune.go]: > what's the purpose of adding these watches? Trying to handle the case that a reference to a layer is added to an image or imagestream while we're in the middle of pruning?

exactly

> It seems unrelated to the fundamental purpose of the PR

Yes, it's unrelated, but the way the code is structured, it's now easy to add that safety mechanism. From the past we know that pruning can take hours to complete and that's way too big window for changes and inconsistencies to happen.
For the sake of simplicity (or rather lesser complexity), I can extract that and move to a follow-up if desired.

File: pkg/oc/admin/prune/imageprune/worker.go (hunk 14 above)
Comment (miminar) [reply pkg/oc/admin/prune/imageprune/worker.go]: fixed

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 13 above)
Comment (bparees) [reply pkg/oc/admin/prune/imageprune/prune.go]: it's fine, just wanted to understand why it was being done here, thanks.

**Comment by bparees** [comment]
/approve
/hold
(hold until 3.11)

**Comment by miminar** [comment]
Rebased

**Comment by bparees** [comment]
@deads2k any concerns w/ the bump commit here?
